### PR TITLE
impr: added correct placement permissions for users (FPLA-1213)

### PR DESCRIPTION
**Pull request by swalker125**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-1213

### Change description ###

confidentialPlacements (all docs) - visible by gatekeeper, admin, Judge and LASOLICITOR
placementWithoutPlacementOrder (confidential docs, no placement order) - visible by cafcass
placements (no confidential docs, no placement order) - visible by remaining users such as SOLICITOR

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 1]
```diff
@@ -75,7 +75,7 @@
     @NotNull(message = "You need to add details to applicant")
     @Valid
     private final List<@NotNull(message = "You need to add details to applicant")
-        Element<Applicant>> applicants;
+            Element<Applicant>> applicants;
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: this auto-format seems incorrect, might want to revert it

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 2]
```diff
@@ -154,7 +156,7 @@
     @Valid
     private final List<@NotNull(message = "You need to add details to children") Element<Child>> children1;
     @NotBlank(message = "Enter Familyman case number", groups = {NoticeOfProceedingsGroup.class,
-        ValidateFamilyManCaseNumberGroup.class})
+            ValidateFamilyManCaseNumberGroup.class})
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: see above

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 3]
```diff
@@ -248,7 +250,7 @@ private void setCaseManagementOrder_Judiciary(CaseManagementOrder order) {
     @TimeNotMidnight(message = "Enter a valid end time", groups = EPOEndDateGroup.class)
     @Future(message = "Enter an end date in the future", groups = EPOEndDateGroup.class)
     @TimeRange(message = "Date must be within the next 8 days", groups = EPOEndDateGroup.class,
-        maxDate = @TimeDifference(amount = 8, unit = DAYS))
+            maxDate = @TimeDifference(amount = 8, unit = DAYS))
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: see above

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 4]
```diff
@@ -273,7 +275,7 @@ private void setCaseManagementOrder_Judiciary(CaseManagementOrder order) {
 
     public Optional<Respondent> findRespondent(int seqNo) {
         return isEmpty(getRespondents1()) || getRespondents1().size() <= seqNo
-            ? empty() : Optional.of(getRespondents1().get(seqNo).getValue());
+                ? empty() : Optional.of(getRespondents1().get(seqNo).getValue());
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: see above

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java [hunk 5]
```diff
@@ -38,30 +38,35 @@ public DynamicList getChildrenList(CaseData caseData, Element<Child> selectedChi
...
-            .orElse(Placement.builder()
-                .childId(child.getId())
-                .childName(child.getValue().getParty().getFullName())
-                .build());
+        return findPlacement(getPlacements(caseData), child.getId())
+                .map(Element::getValue)
+                .orElse(Placement.builder()
+                        .childId(child.getId())
+                        .childName(child.getValue().getParty().getFullName())
+                        .build());
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java]: again the formatting looks too indented to me, shouldn't be double tabbed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java [hunk 6]
```diff
@@ -38,30 +38,35 @@ public DynamicList getChildrenList(CaseData caseData, Element<Child> selectedChi
...
-            .ifPresentOrElse(existingPlacement -> {
-                Element<Placement> newPlacement = element(existingPlacement.getId(), placement);
-                placements.remove(existingPlacement);
-                placements.add(newPlacement);
-            }, () -> placements.add(element(placement)));
+                .ifPresentOrElse(existingPlacement -> {
+                    Element<Placement> newPlacement = element(existingPlacement.getId(), placement);
+                    placements.remove(existingPlacement);
+                    placements.add(newPlacement);
+                }, () -> placements.add(element(placement)));
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java]: see above formatting comments

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java [hunk 7]
```diff
@@ -38,30 +38,35 @@ public DynamicList getChildrenList(CaseData caseData, Element<Child> selectedChi
...
 
         return placements;
     }
 
     private static Optional<Element<Placement>> findPlacement(List<Element<Placement>> placements, UUID childId) {
         return placements.stream()
-            .filter(placement -> placement.getValue().getChildId().equals(childId))
-            .findFirst();
+                .filter(placement -> placement.getValue().getChildId().equals(childId))
+                .findFirst();
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/PlacementService.java]: ☝️

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java [hunk 8]
```diff
@@ -42,15 +49,15 @@ void shouldUpdateChildPlacement() {
...
-                "placements", List.of(child2Placement, child1Placement),
-                "placement", child2NewPlacement.getValue(),
-                "childrenList", child2.getId()))
-            .build();
+                .data(Map.of(
+                        "children1", List.of(child1, child2),
+                        "placements", List.of(child2Placement, child1Placement),
+                        "placement", child2NewPlacement.getValue(),
+                        "childrenList", child2.getId()))
+                .build();
```
Comment (joez-f) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: hm looks like your auto formatter is applying double indents when it shouldn't be, might want to reconfigure / double check

File: ccd-definition/CaseField.json [hunk 9]
```diff
@@ -1072,5 +1072,23 @@
...
+    "ID": "confidentialPlacements",
+    "Label": "Child",
+    "FieldType": "Collection",
+    "FieldTypeParameter": "Placement",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "placementWithoutPlacementOrder",
```
[the remark this answers is not in the document]
Comment (swalker125) [reply ccd-definition/CaseField.json]: sure 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 10]
```diff
@@ -82,14 +83,35 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
         Element<Child> child = placementService.getChild(caseData, childId);
 
         Placement placement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
-            .setChild(child);
+                .setChild(child);
+
+        // add placement with confidential details but no placementOrder.
+        if (placement.hasPlacementOrder()) {
```
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: yep, so I don't actually need the if 👍

[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: also don't care about the has placement order method in Placement

[the remark this answers is not in the document]
Comment (joez-f) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: 👍

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java [hunk 11]
```diff
@@ -60,10 +67,66 @@ void shouldUpdateChildPlacement() {
...
 
-        assertThat(updatedCaseDetails.get("placements")).isEqualTo(List.of(
+        List<String> placements = List.of("placements", "placementsWithoutPlacementOrder", "confidentialPlacements");
+
+        placements.forEach(type -> assertThat(updatedCaseDetails.get(type)).isEqualTo(List.of(
             expectedPlacement(child1Placement, child1Application),
-            expectedPlacement(child2NewPlacement, child2NewApplication)
-        ));
+            expectedPlacement(child2NewPlacement, child2NewApplication)))
+
```
Comment (joez-f) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: 👀

Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/PlacementAboutToSubmitEventControllerTest.java]: I was testing you, well done

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.reform.fpl.model.PlacementOrderAndNotices.PlacementOrderAndNoticesType.NOTICE_OF_HEARING;
+import static uk.gov.hmcts.reform.fpl.model.PlacementOrderAndNotices.PlacementOrderAndNoticesType.PLACEMENT_ORDER;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.unwrapElements;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.wrapElements;
+
+class PlacementTest {
+
+    @Test
+    void shouldReturnEmptyListWhenOnlyPlacementOrderIsPresentInListOfOrderAndNotices() {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java]: wrong method name, it returns Placement object and not list
maybe shouldReturnPlacementWithEmptyList...

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static uk.gov.hmcts.reform.fpl.model.PlacementOrderAndNotices.PlacementOrderAndNoticesType.NOTICE_OF_HEARING;
+import static uk.gov.hmcts.reform.fpl.model.PlacementOrderAndNotices.PlacementOrderAndNoticesType.PLACEMENT_ORDER;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.unwrapElements;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.wrapElements;
+
+class PlacementTest {
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java]: maybe test for removeConfidentialDocuments

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Test
+    void shouldReturnEmptyListWhenOnlyPlacementOrderIsPresentInListOfOrderAndNotices() {
+        Placement placement = Placement.builder()
+                .orderAndNotices(wrapElements(placementOrderAndNoticeOfType(PLACEMENT_ORDER)))
+                .build();
+
+        assertThat(unwrapElements(placement.removePlacementOrder().getOrderAndNotices())).isEmpty();
```
Comment (tomaszpowroznik) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java]: this is not really correct test, it asserts that ordersAndNotices is right, but it does not check that rest of fields are preserved. Probably it should also verify that original object is untouched as this is quite critical here
Maybe:
 DocumentReference placementApplication = TestDataHelper.testDocument();

            Placement placement = Placement.builder()
                .application(placementApplication)
                .orderAndNotices(wrapElements(placementOrderAndNoticeOfType(PLACEMENT_ORDER)))
                .build();

            Placement expectedPlacement = Placement.builder()
                .application(placementApplication)
                .build();

            Placement actualPlacement = placement.removePlacementOrder();

            assertThat(actualPlacement).isEqualTo(expectedPlacement);

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 15]
```diff
@@ -84,14 +85,22 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
         Placement placement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
             .setChild(child);
 
-        caseProperties.put("placements", placementService.setPlacement(caseData, placement));
+        caseProperties.put("confidentialPlacements", placementService.setPlacement(placement, caseData.getConfidentialPlacements()));
+
+        caseProperties.put("placementsWithoutPlacementOrder", placementService.setPlacement(placement.removePlacementOrder(), caseData.getPlacementsWithoutPlacementOrder()));
+
+        caseProperties.put("placements", placementService.setPlacement(removeDocuments(placement), caseData.getPlacements()));
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: looks that it is not tested

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java [hunk 16]
```diff
@@ -84,14 +85,22 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
         Placement placement = mapper.convertValue(caseDetails.getData().get("placement"), Placement.class)
             .setChild(child);
 
-        caseProperties.put("placements", placementService.setPlacement(caseData, placement));
+        caseProperties.put("confidentialPlacements", placementService.setPlacement(placement, caseData.getConfidentialPlacements()));
```
Comment (tomaszpowroznik) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: placement should be set only once

File: service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java (hunk 14 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/model/PlacementTest.java]: this is a much better test 👍 , I shall follow for confidentialDoc removal test

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java (hunk 16 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java]: Replied to you on slack, I am going to address this in separate PR if you are ok with that?

**Pull request merged**